versiondb/client/changeset.go

```diff
@@ -0,0 +1,322 @@
+	var buf [4]byte
+	if _, err := io.ReadFull(reader, buf[:3]); err != nil {
+		return nil, err
+	}
+
+	var pair iavl.KVPair
+	pair.Delete = (buf[0] == 1)
+
+	keyLen := binary.LittleEndian.Uint16(buf[1:])
+	pair.Key = make([]byte, keyLen)
+	if _, err := io.ReadFull(reader, pair.Key); err != nil {
+		return nil, err
+	}
+
+	if !pair.Delete {
+		if _, err := io.ReadFull(reader, buf[:]); err != nil {
+			return nil, err
+		}
+		valueLen := binary.LittleEndian.Uint32(buf[:])
+
+		pair.Value = make([]byte, valueLen)
+		if _, err := io.ReadFull(reader, pair.Value); err != nil {
+			return nil, err
+		}
+	}
+
+	return &pair, nil
+}
+
+// openChangeSetFile opens change set file,
+// it handles compressed files automatically,
+// also supports special name "-" to specify stdin.
+func openChangeSetFile(fileName string) (io.ReadCloser, error) {
+	if fileName == "-" {
+		return os.Stdin, nil
+	}
+	var reader io.Reader
+	fp, err := os.Open(fileName)
+	if err != nil {
+		return nil, err
+	}
+	switch {
+	case strings.HasSuffix(fileName, ZlibFileSuffix):
+		zreader, err := zlib.NewReader(fp)
+		if err != nil {
+			_ = fp.Close()
+			return nil, err
+		}
+		reader = bufio.NewReader(zreader)
+	case strings.HasSuffix(fileName, SnappyFileSuffix):
+		reader = snappy.NewReader(fp)
+	default:
+		reader = bufio.NewReader(fp)
+	}
+	return WrapReader(reader, fp), nil
+}
+
+// withChangeSetFile opens change set file and pass the reader to callback,
+// it closes the file immediatelly after callback returns.
+func withChangeSetFile(fileName string, fn func(io.Reader) error) error {
+	reader, err := openChangeSetFile(fileName)
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+	return fn(reader)
+}
+
+// IterateChangeSets iterate the change set files,
+func IterateChangeSets(
+	reader io.Reader,
+	fn func(version int64, changeSet *iavl.ChangeSet) (bool, error),
+) (int64, error) {
+	var (
+		cont               = true
+		lastCompleteOffset int64
+		version, offset    int64
+		changeSet          *iavl.ChangeSet
+		err                error
+	)
+
+	for cont {
+		version, offset, changeSet, err = ReadChangeSet(reader, true)
+		if err != nil {
+			break
+		}
+		cont, err = fn(version, changeSet)
+		if err != nil {
+			break
+		}
+
+		lastCompleteOffset += offset
+	}
+
+	if err == io.EOF {
+		// it's not easy to distinguish normal EOF or unexpected EOF,
+		// there could be potential corrupted end of file and the err is a normal io.EOF here,
+		// user should verify the change set files in advance, using the verify command.
+		err = nil
+	}
+
+	return lastCompleteOffset, err
+}
+
+// IterateVersions iterate the version numbers in change set files, skipping the change set payloads.
+func IterateVersions(
+	reader io.Reader,
+	fn func(version int64) (bool, error),
+) (int64, error) {
+	var (
+		cont               = true
+		lastCompleteOffset int64
+		version, offset    int64
+		err                error
+	)
+
+	for cont {
```

Comment:
the same question here.

Reply:
it's currently only used in cli, user can Ctl+C directly, I think for enough for now?